packages/extension-mention/src/mention.ts

```diff
@@ -240,30 +265,63 @@ export const Mention = Node.create<MentionOptions>({
             return false
           }
 
+          // Store node and position for later use
+          let mentionNode: any = null
+          let mentionPos = 0
+
           state.doc.nodesBetween(anchor - 1, anchor, (node, pos) => {
             if (node.type.name === this.name) {
               isMention = true
-              tr.insertText(
-                this.options.deleteTriggerWithBackspace ? '' : this.options.suggestion.char || '',
-                pos,
-                pos + node.nodeSize,
-              )
-
+              mentionNode = node
+              mentionPos = pos
               return false
             }
           })
 
+          if (isMention) {
+            tr.insertText(
+              this.options.deleteTriggerWithBackspace ? '' : mentionNode.attrs.mentionSuggestionChar,
+              mentionPos,
+              mentionPos + mentionNode.nodeSize,
+            )
+          }
+
           return isMention
         }),
     }
   },
 
   addProseMirrorPlugins() {
-    return [
-      Suggestion({
+    // Create a plugin for each suggestion configuration
+    return this.storage.suggestions.map(suggestion => {
+      return Suggestion({
+        ...suggestion,
+      })
+    })
+  },
+
+  onBeforeCreate() {
+    this.storage.suggestions = (
+      this.options.suggestions.length ? this.options.suggestions : [this.options.suggestion]
+    ).map(suggestion => {
+      return getSuggestionOptions({
         editor: this.editor,
-        ...this.options.suggestion,
-      }),
-    ]
+        overrideSuggestionOptions: suggestion,
+        extensionName: this.name,
+        char: suggestion.char,
+      })
+    })
+
+    this.storage.getSuggestionFromChar = char => {
```

Comment:
No problem to do something like this but it brings me back to an idea I had before. It feels weird to store internal functions in storage as I'd expect storage to contain data useful for the users.

In the future in another breaking change we could maybe change the API of the `Extension.create` call to accept a callback function like this:

```js
// we could maybe even have editor/other context related things as arguments
const MyNode = Node.create((editor) => {
  function internalHelperFunction() {

  }

  const internalVar = "my-internal-var"

  return {
    // ... your extension config as usual
  }
})
```

what are your thoughts?

Reply:
But no need to do anything right now

Reply:
I agree with this. Sometimes a helper function can help you reuse duplicated code.

Reply:
In general having an enclosed scope for an extension can be really helpful. We could make this change retroactive by allowing both a callback function & the old way of creating extensions.

Reply:
Created a PR for this:
https://github.com/ueberdosis/tiptap/pull/6301

Maybe we can merge this and use this instead of using storage?